### PR TITLE
3.0: Implement a CreateCluster API similar to the ImageBuilder API

### PR DESCRIPTION
**Pull request by ermanno**
### Notes
This patch implements a CreateCluster API similar to the ImageBuilder API, notably:
1. The `Cluster` business object throws granular exceptions using the subclass approach adopted for the ImageBuilder business object.
2. As for the Custom Image APIs, we use a decorator to convert them to the appropriate exception types in the controller. I have made some base exception classes for the main exception types (`LimitExceeded`, `BadRequest` and `Conflict` exceptions), in order to simplify the logic of the exception converters in the `ImageBuilder` and `Cluster` models, and to abstract the decorator so as to reuse it in both the Cluster and Custom Image APIs controllers.
3. We introduce a validate operation for the create request. This is used both by the dryrun operation in the controller, and by the create operation (in particular, we always validate that there is no cluster with the same name prior to creating a cluster).
4. We add the support for named validators for the Buildimage API.

### Tests
1. Unit tests
2. Create cluster
```
curl --location --request POST '{{endpoint}}/v3/clusters?suppressValidators=ALL&validationFailureLevel=WARNING&dryrun=false&rollbackOnFailure=true' \
--header 'Content-Type: application/json' \
--header 'Authorization: true' \
--data-raw '{
    "clusterConfiguration": {{base64encodedconfig}},
    "name": "test-cluster-4",
    "region": "eu-west-1"
}'

{
    "cluster": {
        "cloudformationStackArn": {{stackArn}},
        "cloudformationStackStatus": "CREATE_IN_PROGRESS",
        "clusterName": "test-cluster-4",
        "clusterStatus": "CREATE_IN_PROGRESS",
        "region": "eu-west-1",
        "version": "3.0.0"
    }
}
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/api/controllers/cluster_operations_controller.py [hunk 1]
```diff
@@ -69,22 +78,45 @@ def create_cluster(
...
+    validation_failure_level = validation_failure_level or ValidationLevel.ERROR
+    dryrun = dryrun or False
     create_cluster_request_content = CreateClusterRequestContent.from_dict(create_cluster_request_content)
-    if create_cluster_request_content.cluster_configuration == "invalid":
-        raise CreateClusterBadRequestException(
-            CreateClusterBadRequestExceptionResponseContent(configuration_validation_errors=[], message="invalid")
+
+    # Validate inputs
+    if client_token:
+        raise BadRequestException("clientToken is currently not supported for this operation")
```
[the remark this answers is not in the document]
Comment (ermanno) [reply cli/src/pcluster/api/controllers/cluster_operations_controller.py:89]: I will submit a separate PR where I will remove it from all APIs.

File: cli/src/pcluster/models/cluster.py [hunk 2]
```diff
@@ -289,21 +326,29 @@ def create(
...
+        """Validate a create cluster request."""
+        self._validate_no_existing_stack()
+        self.config, ignored_validation_failures = self._validate_and_parse_config(
+            validator_suppressors, validation_failure_level
+        )
+        return ignored_validation_failures
+
+    def _validate_no_existing_stack(self):
+        if AWSApi.instance().cfn.stack_exists(self.stack_name):
+            raise ConfigValidationError(f"cluster {self.name} already exists")
```
[the remark this answers is not in the document]
Comment (ermanno) [reply cli/src/pcluster/models/cluster.py:349]: Agreed, will return a `BadRequestClusterActionError`.

File: cli/tests/pcluster/api/controllers/test_cluster_operations_controller.py [hunk 3]
```diff
@@ -41,32 +49,360 @@ def cfn_describe_stack_mock_response(edits=None):
...
+            "missing_config",
+            "invalid_region",
+            "missing_name",
+            "invalid_suppress_validators",
+            "invalid_suppress_validators",
+            "invalid_failure_level",
+            "invalid_dryrun",
+            "invalid_rollback",
+            "invalid_config_encoding",
+            # "invalid_config_format",
```
[the remark this answers is not in the document]
Comment (ermanno) [reply cli/tests/pcluster/api/controllers/test_cluster_operations_controller.py:340]: If I let the Marshmallow `ValidationError` propagate I would get the following:

```
{'configurationValidationErrors': [{'level': 'ERROR', 'message': "[('_schema', ['Invalid input type.'])]", 'type': 'ConfigSchemaValidator'}], 'message': 'Invalid cluster configuration'}
```

**Review by demartinofra — APPROVED** [review]
👍🏻 👍🏻 👍🏻

**Pull request merged**